### PR TITLE
Free Foldable and Traverse

### PR DESCRIPTION
**Pull request by i-walker**
Tries to implement this https://github.com/arrow-kt/arrow/issues/720

### CODE REVIEW

File: modules/free/arrow-free-data/src/test/kotlin/arrow/free/FreeTest.kt [hunk 1]
```diff
@@ -60,6 +62,7 @@ class FreeTest : UnitSpec() {
       EqLaws.laws(EQ) { Ops.value(it) },
       MonadLaws.laws(Ops, EQ),
       MonadLaws.laws(Free.monad(), EQ)
+       //FoldableLaws.laws(Free.foldable(), { Free.just<S,Int>(it) }, Eq.any())
```
Comment (raulraja) [inline modules/free/arrow-free-data/src/test/kotlin/arrow/free/FreeTest.kt]: Should this be testing the TraverseLaws?

Comment (i-walker) [reply modules/free/arrow-free-data/src/test/kotlin/arrow/free/FreeTest.kt]: Yes, once I get the code running

File: modules/free/arrow-free/src/main/kotlin/arrow/free/extensions/free.kt [hunk 2]
```diff
@@ -83,3 +86,55 @@ interface FreeEq<F, G, A> : Eq<Kind<FreePartialOf<F>, A>> {
...
+      onPure = { a -> f(a, lb) },
+      onSuspend = { foldRight(lb, f) },
+      onFlatMapped = { _, g: (A) -> Free<F, A> -> foldRight(lb) { a, acc -> g(a).foldRight(acc, f) } }
+    )
+}
+
+@extension
+@undocumented
+interface FreeTraverse<F> : Traverse<FreePartialOf<F>> {
+  val FF: Functor<F>
```
Comment (i-walker) [inline modules/free/arrow-free/src/main/kotlin/arrow/free/extensions/free.kt]: How can I override that element

File: modules/free/arrow-free-data/src/test/kotlin/arrow/free/FreeTest.kt [hunk 3]
```diff
@@ -56,18 +64,21 @@ class FreeTest : UnitSpec() {
...
     val EQ: FreeEq<ForOps, ForId, Int> = Free.eq(IdMonad, idInterpreter)
 
+    @Suppress("UNCHECKED_CAST")
     testLaws(
       EqLaws.laws(EQ) { Ops.value(it) },
       MonadLaws.laws(Ops, EQ),
-      MonadLaws.laws(Free.monad(), EQ)
+      MonadLaws.laws(Free.monad(), EQ),
+      FoldableLaws.laws(Free.foldable(), { Ops.value(it) }, Eq.any()),
+      TraverseLaws.laws(Free.traverse(Id.functor()), Free.functor(), { a -> Ops.value(a) as Kind<FreePartialOf<ForId>, Int> }, Eq.any())
```
Comment (i-walker) [inline modules/free/arrow-free-data/src/test/kotlin/arrow/free/FreeTest.kt]: @nomisRev I am having Type inference issues here is there a way to write down the laws without the UncheckedCast

**Pull request merged**